cdap-client-tests/src/test/java/co/cask/cdap/client/RouteConfigClientTestRun.java

```diff
@@ -0,0 +1,184 @@
+  private ArtifactClient artifactClient;
+  private RouteConfigClient routeConfigClient;
+  private ProgramClient programClient;
+
+  @Before
+  public void setUp() throws Throwable {
+    super.setUp();
+    appClient = new ApplicationClient(clientConfig);
+    artifactClient = new ArtifactClient(clientConfig);
+    routeConfigClient = new RouteConfigClient(clientConfig);
+    programClient = new ProgramClient(clientConfig);
+    // Add AppReturnsArgs artifact
+    addAppReturnsArgsArtifact();
+  }
+
+  @Test
+  public void test() throws Exception {
+    // Deploy AppReturnsArgs of version v1 and start its service
+    String version1 = "v1";
+    ApplicationId app1 = new ApplicationId(NamespaceId.DEFAULT.getNamespace(), AppReturnsArgs.NAME, version1);
+    Map<String, String> propMap1 = deployAppVersionStartService(app1);
+
+    // RouteConfig is empty by default
+    Assert.assertTrue(getRouteConfig().isEmpty());
+
+    // Calls to the non-versioned service endpoint are routed to the only running service of version v1
+    assertArgsEqual(propMap1);
+
+    // Deploy AppReturnsArgs of version v2 and start its service
+    String version2 = "v2";
+    ApplicationId app2 = new ApplicationId(NamespaceId.DEFAULT.getNamespace(), AppReturnsArgs.NAME, version2);
+    Map<String, String> propMap2 = deployAppVersionStartService(app2);
+
+    // Storing and getting RouteConfig with total percentage equal to 100 will succeed
+    storeAndGetValidRouteConfig(ImmutableMap.<String, Integer>of(version1, 10, version2, 90));
+    storeAndGetValidRouteConfig(ImmutableMap.<String, Integer>of(version1, 20, version2, 80));
+    storeAndGetValidRouteConfig(ImmutableMap.<String, Integer>of(version1, 60, version2, 40));
+    
+    // Route all calls to the non-versioned service endpoint are routed to the service of v2
+    storeAndGetValidRouteConfig(ImmutableMap.<String, Integer>of(version1, 0, version2, 100));
+    assertArgsEqual(propMap2);
+
+    // Setting RouteConfig with total percentage not equal to 100 will fail
+    storeInvalidRouteConfig(ImmutableMap.<String, Integer>of(version1, 100, version2, 1), "Percentage");
+    // Setting RouteConfig with total percentage not equal to 100 will fail
+    storeInvalidRouteConfig(ImmutableMap.<String, Integer>of(version1, 98, version2, 1), "Percentage");
+    // Setting RouteConfig with non-existing version will fail
+    storeInvalidRouteConfig(ImmutableMap.<String, Integer>of(version1, 100, "NONEXISTING", 0), "NONEXISTING");
+
+    // Delete RouteConfig
+    routeConfigClient.deleteRouteConfig(NamespaceId.DEFAULT, AppReturnsArgs.NAME, AppReturnsArgs.SERVICE);
+    // RouteConfig is empty after deletion
+    Assert.assertTrue(getRouteConfig().isEmpty());
+
+    // Stop all services
+    programClient.stopAll(Id.Namespace.DEFAULT);
+    // Delete all apps
+    appClient.deleteAll(Id.Namespace.DEFAULT);
+  }
+
+  private Map<String, String> deployAppVersionStartService(ApplicationId app) throws Exception {
```

Comment:
but to avoid duplicate code, this method becomes weird: Why should something that starts a service return the runtime arguments? It seems like runtime arguments should be provided by the caller of the method that starts the service.